[Executorch][llama] Allow custom sdpa op replacement pass to leverage attention mask

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #10286
* __->__ #10285
* #10284
* #10283

Previously we assumed that the custom sdpa always does causal attention.
This diff adds option to this module swap pass to make custom sdpa leverage
attention mask instead of causal.

Differential Revision: [D73222736](https://our.internmc.facebook.com/intern/diff/D73222736/)

kimishpatel committed:
[Executorch][llama] Allow custom sdpa op replacement pass to leverage attention mask

Previously we assumed that the custom sdpa always does causal attention.
This diff adds option to this module swap pass to make custom sdpa leverage
attention mask instead of causal.

Differential Revision: [D73222736](https://our.internmc.facebook.com/intern/diff/D73222736/)

[ghstack-poisoned]

lucylq commented on examples/models/llama/source_transformation/sdpa.py:
```diff
@@ -47,34 +63,54 @@ def forward(
-            0,  # dropout probability. Ignored by the code
-            True,  # is_causal
-        )
+        if self.use_attention_mask:
```
nit, maybe this could be:

if not self.use_attention_mask:
    mask = None
    
output = torch.ops.llama.custom_sdpa(
                q,
                k,
                v,
                input_pos[0].item(),
                mask,  # Attention mask
                0,  # dropout probability. Ignored by the code
                False,  # is_causal
            )

kimishpatel replied:
Note that is_causal is True in one case, while not true in the other

kimishpatel committed:
Update on "[Executorch][llama] Allow custom sdpa op replacement pass to leverage attention mask"


Previously we assumed that the custom sdpa always does causal attention.
This diff adds option to this module swap pass to make custom sdpa leverage
attention mask instead of causal.

Differential Revision: [D73222736](https://our.internmc.facebook.com/intern/diff/D73222736/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][llama] Allow custom sdpa op replacement pass to leverage attention mask"


Previously we assumed that the custom sdpa always does causal attention.
This diff adds option to this module swap pass to make custom sdpa leverage
attention mask instead of causal.

Differential Revision: [D73222736](https://our.internmc.facebook.com/intern/diff/D73222736/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch][llama] Allow custom sdpa op replacement pass to leverage attention mask"


Previously we assumed that the custom sdpa always does causal attention.
This diff adds option to this module swap pass to make custom sdpa leverage
attention mask instead of causal.

Differential Revision: [D73222736](https://our.internmc.facebook.com/intern/diff/D73222736/)

[ghstack-poisoned]
